### DISCUSSION TITLE
Uno app not using Framebuffer display

### DISCUSSION THREAD

**Discussion by ashah7** [unoplatform/uno#17044] [Q&A] [answered]
I'm trying to get one of the sample UNO projects to use the framebuffer on my Raspberry pi 3.  Currently it has two enabled `/dev/fb0 and /dev/fb1.`

I can display a image using fbi to either framebuffer and it shows on the screens
All the debugging has been enabled in the app
[uno-logs-framebuffer.txt](https://github.com/user-attachments/files/15742966/uno-logs-framebuffer.txt)

Attached file shows when using `DISPLAY=:0` app is working fine
the other from using` DISPLAY=:1`, this doesn't exist and silently exits (as expected)

using export `FRAMEBUFFER=/dev/fb1 or 0` also silently exits

Note that when using the FRAMEBUFFER I have disabled all X servers

Anything I'm missing to get this working or help troubleshooting this further?

```

pi@000000007cb01ee3:~ $ sudo fbi -T 2 -d /dev/fb0 -noverbose -a /home/homeradio/LCD-HDMI-MessageRpi3.jpg
using "Noto Sans Mono-16", pixelsize=16.67 file=/usr/share/fonts/truetype/noto/NotoSansMono-Regular.ttf

pi@000000007cb01ee3:~ $ export FRAMEBUFFER=/dev/fb0
pi@000000007cb01ee3:~ $ dotnet Menu/Menu.dll


pi@000000007cb01ee3:~ $ dotnet --info
.NET SDK:
 Version:           8.0.301
 Commit:            1a0e9c0300
 Workload version:  8.0.300-manifests.f6879a9a
 MSBuild version:   17.10.4+10fbfbf2e

Runtime Environment:
 OS Name:     debian
 OS Version:  12
 OS Platform: Linux
 RID:         linux-arm64
 Base Path:   /opt/dotnet/sdk/8.0.301/

.NET workloads installed:
There are no installed workloads to display.

Host:
  Version:      8.0.6
  Architecture: arm64
  Commit:       3b8b000a0e

.NET SDKs installed:
  8.0.301 [/opt/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 8.0.6 [/opt/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 8.0.6 [/opt/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download

pi@000000007cb01ee3:~ $ /opt/dotnet/tools/uno-check
You must install .NET to run this application.

App: /opt/dotnet/tools/uno-check
Architecture: arm64
App host version: 8.0.6
.NET location: Not found

Learn more:
https://aka.ms/dotnet/app-launch-failed

Download the .NET runtime:
https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=arm64&rid=linux-arm64&os=debian.12&apphost_version=8.0.6 


```
```

pi@000000007cb01ee3:~ $ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
NAME="Debian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

pi@000000007cb01ee3:~ $ uname -r
6.6.31-v8+

pi@000000007cb01ee3:~ $ uname -a
Linux 000000007cb01ee3 6.6.31-v8+ #1766 SMP PREEMPT Fri May 24 12:14:10 BST 2024 aarch64 GNU/Linux
```

`Model           : Raspberry Pi 3 Model B Rev 1.2`

**Comment by jeromelaban** [comment]
Thanks for the question. Can you provide again the full linux distribution information? We're testing regularly on RPi3s and it's generally working fine.

**Reply by ashah7** [reply]
original answer appended with the information

**Reply by agneszitte** [reply]
@jeromelaban, I'm adding the details of the previous discussion with @ashah7 from [Discord](https://discord.com/channels/1182775715242967050/1182775715729526908/1248654773377302661) for reference below:
<details>

**<summary>Click here to expand to see the details:</summary>**


> AShah — 06/05/2024 4:29 AM
> Hi - I'm trying to get one of the sample projects working on a raspberry pi 3b but when running it just gives me a black screen.  When using my local windows machine to display it works fine.  I'm using Raspbian bookworm Lite  
> I've had a look at the below and followed the instructions
> https://github.com/unoplatform/uno/discussions/11833
> https://github.com/unoplatform/uno/issues/4524
> https://platform.uno/docs/articles/get-started-with-linux.html?tabs=ubuntu1804#setting-up-for-linux

> Jerome Laban — 06/05/2024 10:04 AM
> Are there any errors showing in the logs? Make sure that the InitializeLogging method is called from the Program.cs file

> AShah — 06/05/2024 1:01 PM
> okay will try this and report

> AShah — 06/05/2024 5:47 PM
> This seems to work fine on the full Raspbian desktop version but not Lite
> Is there a guide or something for support for Raspbian Lite?

> Jerome Laban — 06/05/2024 8:29 PM
> There is no guide that I know of. If you can provide any logging, this will help troubleshoot.

> AShah — Yesterday at 12:10 PM
> manage to find a solution, it seems like it it won't display with standard X11 server running, I had to use openbox to get it going.
> However even without no X server running according to the docs I should still be able to display using Framebuffer support
> setting the env variable using "export FRAMEBUFFER=/dev/fb1" to display on my LCD doesn't show or give anything in the logs it just silently exits.

> Jerome Laban — Yesterday at 12:16 PM
> That's right. You should still be seeing some messages in the app's logs. Were you able to ensure that InitializeLogging is called?

> AShah — Yesterday at 12:36 PM
> I did try using this 
> https://platform.uno/docs/articles/logging.html#configuring-logging
> but couldn't get it going, I found no reference to program.cs.  I think that guide isn't for newbies so I had trouble following it
> 
> Very crudely I did the below in the App.xaml.cs (don't laugh)
> 
> ```
>  .UseLogging(configure: (context, logBuilder) =>
>  {
>      // Configure log levels for different categories of logging
>      logBuilder
>          .SetMinimumLevel(
>              //context.HostingEnvironment.IsDevelopment() ?
>              //    LogLevel.Information :
>              //    LogLevel.Warning 
>                  
>                 LogLevel.Debug  <---added this
>                  )
> ......
> ```
> 
> After rebuilding it did start giving me more logs when running the app normally but still nothing when trying to display using framebuffer
> 
> below showing my framebuffer is valid I can send an image to it using feh or fbi
> 
> ```
> fbset -fb /dev/fb1
> 
> mode "480x320"
>     geometry 480 320 480 320 32
>     timings 0 0 0 0 0 0 0
>     rgba 8/16,8/8,8/0,0/0
> endmode 
> ```

> Jerome Laban — Yesterday at 1:18 PM
> Right, if you're using the recommended template, the logging doc is this one. We'll make a change to redirect from one to the other: https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Logging/LoggingOverview.html
> You may want to put CoreLogLevel to debug as well 

> AShah — Yesterday at 5:47 PM
> corelevel is set to debug and so is everything else down the list but still nothing appearing, running fbi I can see the image on my display so the fb is working
> X is all disabled
> 
> ```
> pi@000000007cb01ee3:~ $ sudo fbi -T 2 -d /dev/fb0 -noverbose -a /home/homeradio/LCD-HDMI-MessageRpi3.jpg
> using "Noto Sans Mono-16", pixelsize=16.67 file=/usr/share/fonts/truetype/noto/NotoSansMono-Regular.ttf
> 
> pi@000000007cb01ee3:~ $ export FRAMEBUFFER=/dev/fb0
> pi@000000007cb01ee3:~ $ dotnet Menu/Menu.dll
> 
> 
> pi@000000007cb01ee3:~ $ dotnet --info
> .NET SDK:
>  Version:           8.0.301
>  Commit:            1a0e9c0300
>  Workload version:  8.0.300-manifests.f6879a9a
>  MSBuild version:   17.10.4+10fbfbf2e
> 
> Runtime Environment:
>  OS Name:     debian
>  OS Version:  12
>  OS Platform: Linux
>  RID:         linux-arm64
>  Base Path:   /opt/dotnet/sdk/8.0.301/
> 
> .NET workloads installed:
> There are no installed workloads to display.
> 
> Host:
>   Version:      8.0.6
>   Architecture: arm64
>   Commit:       3b8b000a0e
> 
> .NET SDKs installed:
>   8.0.301 [/opt/dotnet/sdk]
> 
> .NET runtimes installed:
>   Microsoft.AspNetCore.App 8.0.6 [/opt/dotnet/shared/Microsoft.AspNetCore.App]
>   Microsoft.NETCore.App 8.0.6 [/opt/dotnet/shared/Microsoft.NETCore.App]
> 
> Other architectures found:
>   None
> 
> Environment variables:
>   Not set
> 
> global.json file:
>   Not found
> 
> Learn more:
>   https://aka.ms/dotnet/info
> 
> Download .NET:
>   https://aka.ms/dotnet/download
> 
> pi@000000007cb01ee3:~ $ /opt/dotnet/tools/uno-check
> You must install .NET to run this application.
> 
> App: /opt/dotnet/tools/uno-check
> Architecture: arm64
> App host version: 8.0.6
> .NET location: Not found
> 
> Learn more:
> https://aka.ms/dotnet/app-launch-failed
> 
> Download the .NET runtime:
> https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=arm64&rid=linux-arm64&os=debian.12&apphost_version=8.0.6
> ```

> Jerome Laban — Yesterday at 8:52 PM
> Does it terminate with a coredump or does it just show nothing?

> AShah — Yesterday at 10:52 PM
> just silently fails - see attached for trying to go framebuffer (with no X server) and then DISPLAY=:0
> DISPLAY=:1 doesnt exist and it silently fails as well which I expect. 
> Note that in the dotnet -info output there are no environment variables set, I assume this is not where the FRAMEBUFFER needs to be set but just in the OS
> Attachment file:
> [uno-logs-framebuffer](https://cdn.discordapp.com/attachments/1182775715729526908/1248469529227563038/uno-logs-framebuffer?ex=6664701b&is=66631e9b&hm=508a373f712c3f801ffd4b093b91689530fc28b9395ab4b6b90ab945e7464909&) 29.98 KB

> Jerome Laban — Yesterday at 11:01 PM
> Very interesting. This may point to some native dependency that fails, somehow. Are you able to run with gdb to see if something stands out?

> AShah — Today at 8:19 AM
> never used gdb, but running a google does it even work at the OS level, seems to be used for debugging C and C++ scripts ?
> I may have to come back to this took an age just to get to this point
> I'm going to start putting an interface and application together

> Jerome Laban — Today at 10:27 AM
> This type of crash is generally native, so gdb can help.

</details>

**Reply by jeromelaban** [reply]
@ashah7 I've just installed the raspbian bookwork lite, following the .NET SDK install. This is what I have:

<details><summary>Details</summary>
<p>

```
$ cat /proc/cpuinfo | grep Model
Model : Raspberry Pi 3 Model B Rev 1.2
$ uname -r
6.6.31+rpt-rpi-v8
$ uname -a
Linux raspberrypi 6.6.31+rpt-rpi-v8 #1 SMP PREEMPT Debian 1:6.6.31-1+rpt1 (2024-05-29) aarch64 GNU/Linux
$ $  dotnet --info
.NET SDK:
 Version:           8.0.301
 Commit:            1a0e9c0300
 Workload version:  8.0.300-manifests.f6879a9a
 MSBuild version:   17.10.4+10fbfbf2e

Runtime Environment:
 OS Name:     debian
 OS Version:  12
 OS Platform: Linux
 RID:         linux-arm64
 Base Path:   /home/jay/.dotnet/sdk/8.0.301/

.NET workloads installed:
There are no installed workloads to display.

Host:
  Version:      8.0.6
  Architecture: arm64
  Commit:       3b8b000a0e

.NET SDKs installed:
  8.0.301 [/home/jay/.dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 8.0.6 [/home/jay/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 8.0.6 [/home/jay/.dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  Not set

global.json file:
  Not found
```

</p>
</details> 

The only difference here is that libinput is not installed by default on the lite version.  We're updating the docs but you can add with `apt install libinput10`. I got the app to run with both the default (`dotnet new unoapp -o test01`) and recommended (`dotnet new unoapp --preset recommended -o test02`) templates, then `dotnet run -f net8.0-desktop`.

Can you tell how you built the app in the first place?

Also, it would be using to test with gdb (it's installed by default on the lite build):
1. Go the folder where your `Menu.dll` is
2. Launch `gdb`:
   ```
   gdb dotnet
   ```
3. Once started, type:
   ```
   r Menu.dll
   ```
4. Wait a bit and paste in this thread what you see.

Thanks!

**Reply by ashah7** [reply]
I downloaded the sample from VS 2022 built and copied over to the pi, this runs fine on xserver (using openbox) on the pi or on a redirect to my local machine over ssh

I did try to build like in your commands the test01 project, it came up with the below errors:
```

pi@000000007cb01ee3:~ $ dotnet new unoapp -o test01
The template "Uno Platform App" was created successfully.


pi@000000007cb01ee3:~ $ dotnet new unoapp -o test01
The template "Uno Platform App" was created successfully.

Processing post-creation actions...
Restoring /home/pi/test01/test01/test01.csproj:
  Determining projects to restore...
/opt/dotnet/sdk/8.0.301/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.ImportWorkloads.targets(38,5): error NETSDK1147: To build this project, the following workloads must be installed: wasi-experimental [/home/pi/test01/test01/test01.csproj::TargetFramework=net8.0-android]
/opt/dotnet/sdk/8.0.301/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.ImportWorkloads.targets(38,5): error NETSDK1147: To install these workloads, run the following command: dotnet workload restore [/home/pi/test01/test01/test01.csproj::TargetFramework=net8.0-android]
Restore failed.
Restoring /home/pi/test01/Directory.Packages.props:
Restore succeeded.
Post action failed.
Manual instructions: Run 'dotnet restore'
```


```
pi@000000007cb01ee3:~/AirMyPrayerMenu $ gdb dotnet
GNU gdb (Debian 13.1-3) 13.1
Copyright (C) 2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "aarch64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from dotnet...
(No debugging symbols found in dotnet)
(gdb) r AirMyPrayerMenu.dll
Starting program: /usr/bin/dotnet AirMyPrayerMenu.dll
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fb727f140 (LWP 1586430)]
[New Thread 0x7fb6a6f140 (LWP 1586431)]
[New Thread 0x7fb624f140 (LWP 1586432)]
[New Thread 0x7fb5a3f140 (LWP 1586433)]
[New Thread 0x7fb5100140 (LWP 1586434)]
[New Thread 0x7fa7278140 (LWP 1586435)]
[New Thread 0x7f09e55140 (LWP 1586436)]
[Thread 0x7f09e55140 (LWP 1586436) exited]
[Thread 0x7fa7278140 (LWP 1586435) exited]
[Thread 0x7fb5100140 (LWP 1586434) exited]
[Thread 0x7fb5a3f140 (LWP 1586433) exited]
[Thread 0x7fb6a6f140 (LWP 1586431) exited]
[Thread 0x7fb727f140 (LWP 1586430) exited]
[Thread 0x7ff7fbc040 (LWP 1586428) exited]
[Thread 0x7fb624f140 (LWP 1586432) exited]
[New process 1586428]
[Inferior 1 (process 1586428) exited normally]
(gdb)

```

**Reply by jeromelaban** [reply]
Thanks. In this new app that was created, can you enable again the logging, by setting [this line](https://github.com/unoplatform/uno.templates/blob/adae1be3e850dc74e16b6742be07ee5711ae2514/src/Uno.Templates/content/unoapp/MyExtensionsApp.1/App.blank.cs#L140) and [this line](https://github.com/unoplatform/uno.templates/blob/adae1be3e850dc74e16b6742be07ee5711ae2514/src/Uno.Templates/content/unoapp/MyExtensionsApp.1/App.blank.cs#L137) to Debug ? Then try again with gdb, we'll see if anything else shows up.

Also make sure that [this line](https://github.com/unoplatform/uno.templates/blob/adae1be3e850dc74e16b6742be07ee5711ae2514/src/Uno.Templates/content/unoapp/MyExtensionsApp.1/Platforms/Desktop/Program.cs#L12) is available in your `program.cs`. If not, make sure to run `dotnet new update` to get the latest templates, then recreate a test project or add the line manually.

Thanks!

<details><summary>For the record, this is what I see with a recommended template, where the app runs properly: </summary>
<p>

```
$ gdb -- dotnet
GNU gdb (Debian 13.1-3) 13.1
Copyright (C) 2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "aarch64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from dotnet...
(No debugging symbols found in dotnet)
(gdb) r testfbrec.dll
Starting program: /home/jay/.dotnet/dotnet testfbrec.dll
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fb727f140 (LWP 2565)]
[New Thread 0x7fb6a6f140 (LWP 2566)]
[New Thread 0x7fb624f140 (LWP 2567)]
[New Thread 0x7fb5a3f140 (LWP 2568)]
[New Thread 0x7fb5100140 (LWP 2569)]
[New Thread 0x7fa7278140 (LWP 2570)]
[New Thread 0x7f0a13e140 (LWP 2571)]
[New Thread 0x7f0992e140 (LWP 2572)]
[New Thread 0x7f0885f140 (LWP 2574)]
[New Thread 0x7f0906f140 (LWP 2573)]
[New Thread 0x7ef195c140 (LWP 2611)]
[New Thread 0x7ef114c140 (LWP 2612)]
[New Thread 0x7ef093c140 (LWP 2613)]
[New Thread 0x7ef08ec140 (LWP 2614)]
[New Thread 0x7edfc07140 (LWP 2622)]
warn: Uno.Extensions.Storage.FileStorage[0]
      ReadPackageFileAsync - File 'appsettings.keyvaluestorageconfiguration.json' does not exist in package
warn: Uno.Extensions.Storage.FileStorage[0]
      ReadPackageFileAsync - File 'appsettings.appconfig.json' does not exist in package
warn: Uno.Extensions.Storage.FileStorage[0]
      ReadPackageFileAsync - File 'appsettings.localizationconfiguration.json' does not exist in package
warn: Uno.Extensions.Storage.FileStorage[0]
      ReadPackageFileAsync - File 'appsettings.localizationsettings.json' does not exist in package
warn: Uno.Extensions.Storage.FileStorage[0]
      ReadPackageFileAsync - File 'appsettings.navigationconfiguration.json' does not exist in package
warn: Uno.Extensions.Navigation.Navigators.ContentControlNavigator[0]
      Show - Missing view for navigation path ''
[New Thread 0x7edf3f7140 (LWP 2623)]
fail: Microsoft.UI.Xaml.Controls.ScrollViewer[0]
      Zoom-based content alignment is not implemented on this platform.
```

</p>
</details>

**Reply by ashah7** [reply]
thanks - if your talking about the new test01 app I generated then it didn't build see above error, so there's nothing to run for gdb

As for the Menu.dll I believe debug was all enabled on that

**Reply by jeromelaban** [reply]
@ashah7 it did not build because you if you use the default parameters, on a RPi it require more pieces to install to build entirely that you do not need here. 

You can remove all target frameworks from the generated project so that it shows only contains `<TargetFrameworks>net8.0-desktop</TargetFrameworks>`, it should build then. For the debugging, it seems that it was not enabled, it would have otherwise shown when starting the app.

Let's try to make a default blank app run first.

**Comment by ashah7** [comment, marked as answer]
Managed to find the problem
I was using MobaXterm  as my ssh client, I changed to plain old putty and it worked!

I figured from the below that it still wanted to use X11 by redirecting to the local desktop even though it was disabled

thanks for your help


```
pi@000000007cb01ee3:~/test01/test01/bin/Debug/net8.0-desktop $ dotnet test01.dll
dbug: Uno.UI.Runtime.Skia.SkiaHostBuilder[0]
      Using host builder Uno.UI.Runtime.Skia.X11HostBuilder
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get remote control server port from the IDE for endpoint 192.168.0.168.
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get remote control server port from the IDE for endpoint [fe80::78ea:6d8c:d16b:ae27].
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get any remote control server endpoint from the IDE.
dbug: Uno.UI.Xaml.Core.VisualTree[0]
      Visual Tree was not found.
dbug: Uno.UI.Xaml.Core.VisualTree[0]
      Visual Tree was not found.
dbug: Uno.UI.Xaml.Core.VisualTree[0]
      Visual Tree was not found.
dbug: Uno.UI.Xaml.Core.VisualTree[0]
      Visual Tree was not found.
fail: Uno.WinUI.Runtime.Skia.X11.X11XamlRootHost[0]
      XLIB ERROR: Cannot connect to X server
dbug: Uno.WinUI.Runtime.Skia.X11.X11ApplicationHost[0]
      X11ApplicationHost is exiting
```

**Reply by jeromelaban** [reply]
> I was using MobaXterm as my ssh client, I changed to plain old putty and it worked!

This is great news! If you specify a `DISPLAY` environment variable, the framebuffer more will not be used and X11 will take over:

https://github.com/unoplatform/uno/blob/2e81c9f74478ad380012d7572a87bec2887eb379/src/Uno.UI.Runtime.Skia.X11/Builder/X11HostBuilder.cs#L19

You would need to unset this variable first. You were likely blocked by the fact that the debugging information was not set properly in a previous version of our templates, that `XLIB ERROR: Cannot connect to X server` should have been showing.
